File: LEETCODE/888. 公平的糖果交换.cpp
#include<iostream>
#include<vector>
using namespace std; 
class Solution {
public:
    vector<int> fairCandySwap(vector<int>& A, vector<int>& B) {
        vector<int> ans;
        int lenA = A.size();
	int lenB = B.size();
	
	int sumA=0,sumB=0;
	for(int i=0;i<lenA;i++){
		sumA+=A[i];
	}
	for(int i=0;i<lenB;i++){
		sumB+=B[i];
	}
	
	for(int i=0;i<lenA;i++){
		for(int j=0;j<lenB;j++){
			if(sumA-A[i]+B[j]==sumB-B[j]+A[i]){
				ans.push_back(A[i]);
				ans.push_back(B[j]);
                
				return ans;
			}
		}
	} 
    }
};
int main(){
	vector<int> A,B,ans;
	A.push_back(1);
	A.push_back(2);
	A.push_back(5);
	B.push_back(2);
	B.push_back(4); 
	Solution solution;
	ans = solution.fairCandySwap(A,B);
	cout<<ans[0]<<","<<ans[1]<<endl;
	
	return 0;
}

File: LEETCODE/888. 公平的糖果交换.java
class Solution {
    public int[] fairCandySwap(int[] A, int[] B) {
        // 数组 A 中数字的总和
        int sumA = Arrays.stream(A).sum();
        // 数组 B 中数字的总和
        int sumB = Arrays.stream(B).sum();
        
        // 对一个可行解(x,y)，他们的差值应该为 delta。（公式推导见 ref）
        int delta = (sumA - sumB) / 2;
        // 为了快速查询 A 中是否存在某个数字，先把数字全部存入哈希表
        // 然后再遍历 B，在哈希表中查询是否有差值为 delta 的数字
        Set<Integer> set = new HashSet<Integer>();

        // 遍历 A，存入哈希表
        // set 会自动去重
        for (int num : A) {
            set.add(num);
        }

        // 一个可行解
        int[] ans = new int[2];

        // 遍历 B 中的数字 y，查找哈希表中是否存在一个数字 x 与 y 差值为 delta
        for (int y : B) {
            // 先将 y 加上差值 delta，假设一个 x
            int x = y + delta;
            // 再判断哈希表里是否存在
            if (set.contains(x)) {
                // 若存在，找到一个可行解，886
                ans[0] = x;
                ans[1] = y;
                break;
            }
        }
        return ans;
    }
}

// ref：https://leetcode-cn.com/problems/fair-candy-swap/solution/gong-ping-de-tang-guo-jiao-huan-by-leetc-tlam/

